### ISSUE TITLE
build minio server failed on illumos based platform

### ISSUE THREAD

**Issue by ikozhukhov** [minio/minio#6760]
I have DilOS - it is illumos based platform

try to build minio server and can see issues:

```
root@z165:~# go get -u github.com/minio/minio
# github.com/minio/minio/cmd/http
go/src/github.com/minio/minio/cmd/http/listener.go:398:15: undefined: listen
go/src/github.com/minio/minio/cmd/http/listener.go:399:16: undefined: fallbackListen
```
could you please let me know - how can i try to fix it ?
illumos platform detect as solaris/amd64

**Comment by harshavardhana** [comment]
Should be easy to fix we need to add ``// +build solaris`` file

**Comment by ikozhukhov** [comment]
what file of files?

**Comment by harshavardhana** [comment]
https://github.com/minio/minio/blob/master/cmd/http/listen_others.go in this file at the top just add

```
// +build windows plan9 solaris
```

**Comment by ikozhukhov** [comment]
it failed:
```
root@z165:~/go/src/github.com/minio/minio# go build
# github.com/minio/minio/cmd
cmd/posix-list-dir_unix.go:61:6: undefined: isEmptyDirent
cmd/posix-list-dir_unix.go:71:16: dirent.Type undefined (type *syscall.Dirent has no field or method Type)
cmd/posix-list-dir_unix.go:72:8: undefined: syscall.DT_DIR
```

**Comment by harshavardhana** [comment]
Remove the `solaris` entry from `_unix.go` file and instead add it to https://github.com/minio/minio/blob/master/cmd/posix-list-dir_plan9.go

**Comment by ikozhukhov** [comment]
```
root@z165:~/go/src/github.com/minio/minio# g diff
diff --git a/cmd/http/listen_others.go b/cmd/http/listen_others.go
index 5216cc41..76812cb8 100644
--- a/cmd/http/listen_others.go
+++ b/cmd/http/listen_others.go
@@ -1,4 +1,4 @@
-// +build windows plan9
+// +build windows plan9 solaris

 /*
  * Minio Cloud Storage, (C) 2018 Minio, Inc.
diff --git a/cmd/posix-list-dir_plan9.go b/cmd/posix-list-dir_plan9.go
index 14b8786d..7fecad20 100644
--- a/cmd/posix-list-dir_plan9.go
+++ b/cmd/posix-list-dir_plan9.go
@@ -1,4 +1,4 @@
-// +build plan9
+// +build plan9 solaris

 /*
  * Minio Cloud Storage, (C) 2016, 2017, 2018 Minio, Inc.
diff --git a/cmd/posix-list-dir_unix.go b/cmd/posix-list-dir_unix.go
index 14d7e409..50dffca7 100644
--- a/cmd/posix-list-dir_unix.go
+++ b/cmd/posix-list-dir_unix.go
@@ -1,4 +1,4 @@
-// +build darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build darwin dragonfly freebsd linux nacl netbsd openbsd

 /*
  * Minio Cloud Storage, (C) 2016, 2017, 2018 Minio, Inc.
```
build failed with:
```
root@z165:~/go/src/github.com/minio/minio# go build
# github.com/minio/minio/cmd
cmd/certs.go:73:14: undefined: readDir
cmd/disk-cache-fs.go:166:20: undefined: readDir
cmd/disk-cache.go:385:19: undefined: readDir
cmd/disk-usage.go:41:18: undefined: readDir
cmd/fs-v1-multipart.go:90:18: undefined: readDir
cmd/fs-v1-multipart.go:149:20: undefined: readDir
cmd/fs-v1-multipart.go:380:18: undefined: readDir
cmd/fs-v1-multipart.go:521:18: undefined: readDir
cmd/fs-v1-multipart.go:757:20: undefined: readDir
cmd/fs-v1-multipart.go:762:23: undefined: readDir
cmd/fs-v1-multipart.go:762:23: too many errors
```

**Comment by ikozhukhov** [comment]
fine, just fixed it by:
```
cp posix-list-dir_plan9.go posix-list-dir_solaris.go
```
and update to build only on solaris platform
will try to test final binary
thanks for your help with build!

**Comment by harshavardhana** [comment]
Please send a PR we will merge it.

**Comment by ikozhukhov** [comment]
will do it after tests.

**Comment by ikozhukhov** [comment]
We are using bitbucket for DilOS and I did mirror to: https://bitbucket.org/dilos/minio
special for DilOS we have separate fork for build and fixes.

You can find fixed build at:
https://bitbucket.org/dilos/dilos-minio/commits/155df0233019f25897c8fa49d0ed3e7cda320142

we are using DEB packages and i'll update repo with build of minio as package.
i can let you know with update and if it will be applicable, you can merge it too.

**Comment by harshavardhana** [comment]
@ikozhukhov would you be able to send the PR?

**Comment by ikozhukhov** [comment]
will try to do it on github
could you interested in my updates for DEB package build based on Debian ?
also, i can see failed tests and i'd like to to the best way to work on it:
https://paste.dilos.org/?4a1fa6c8689bd8fe#2llgrft/tAUyAWndsCRpMBXy5Sf0Sipb7rVZ7sam/Mw=
is it can be separate issue per failed test or i can open one?

**Comment by ikozhukhov** [comment]
```
=== RUN   TestFSDeletes
--- FAIL: TestFSDeletes (0.00s)
        fs-v1-helpers_test.go:347: Test case 6: Expected: "%!s(<nil>)", got: "remove /tmp/minio-638393463/success-vol/not-empty: file exists"
=== RUN   TestFSRemoves
--- FAIL: TestFSRemoves (0.00s)
        fs-v1-helpers_test.go:485: remove /tmp/minio-739233642/success-vol: file exists

=== RUN   TestGetObjectPermissionDenied
--- FAIL: TestGetObjectPermissionDenied (0.16s)
        object-api-getobject_test.go:273: Test 1: XL: Expected to fail with <ERROR> "Prefix access is denied: ompcse2c1vv14ez6vktxpatg9uwusk0as2mmjxvty01nbei1qcivsjqhesnn/test-object1", but passed instead.
        object-api-getobject_test.go:278: Test 2: XL: Expected to fail with error "Prefix access is denied: ompcse2c1vv14ez6vktxpatg9uwusk0as2mmjxvty01nbei1qcivsjqhesnn/dir/test-object2", but instead failed with error "Object not found: ompcse2c1vv14ez6vktxpatg9uwusk0as2mmjxvty01nbei1qcivsjqhesnn#dir/test-object2" instead.
        object-api-getobject_test.go:278: Test 3: XL: Expected to fail with error "Prefix access is denied: ompcse2c1vv14ez6vktxpatg9uwusk0as2mmjxvty01nbei1qcivsjqhesnn/test-object3", but instead failed with error "Object not found: ompcse2c1vv14ez6vktxpatg9uwusk0as2mmjxvty01nbei1qcivsjqhesnn#test-object3" instead.

=== RUN   TestPosixMakeVol
--- FAIL: TestPosixMakeVol (0.00s)
        posix_test.go:468: expected: disk access denied, got: %!s(<nil>)
=== RUN   TestPosixDeleteVol
--- FAIL: TestPosixDeleteVol (0.00s)
        posix_test.go:546: TestPosix: 3, expected: volume is not empty, got: remove /tmp/minio-279824838/nonempty-vol: file exists

=== RUN   TestPosixPosixListDir
--- FAIL: TestPosixPosixListDir (0.00s)
        posix_test.go:876: expected: file access denied, got: %!s(<nil>)
=== RUN   TestPosixDeleteFile
--- FAIL: TestPosixDeleteFile (0.00s)
        posix_test.go:1022: expected: file access denied, got: %!s(<nil>)
=== RUN   TestPosixReadFile
--- FAIL: TestPosixReadFile (0.00s)
        posix_test.go:1148: Creating file failed: 4 struct { volume string; fileName string; offset int64; bufSize int; expectedBuf []uint8; expectedErr error }{volume:"success-vol", fileName:"level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003/object000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001", offset:0, bufSize:5, expectedBuf:[]uint8{0x68, 0x65, 0x6c, 0x6c, 0x6f}, expectedErr:error(nil)}, expected: %!s(<nil>), got: open /tmp/minio-286033808/success-vol/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003/object000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001: file name too long
=== RUN   TestPosixReadFileWithVerify
--- PASS: TestPosixReadFileWithVerify (0.00s)
=== RUN   TestPosixAppendFile
--- FAIL: TestPosixAppendFile (0.00s)
        posix_test.go:1388: Case: 10, expected: %!s(<nil>), got: open /tmp/minio-099676994/success-vol/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003/object000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001: file name too long
        posix_test.go:1414: expected: Permission error, got: %!s(<nil>)
=== RUN   TestPosixPrepareFile
--- FAIL: TestPosixPrepareFile (0.00s)
        posix_test.go:1489: Case: 7, expected: %!s(<nil>), got: open /tmp/minio-522472388/success-vol/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003/object000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001: file name too long
        posix_test.go:1515: expected: Permission error, got: %!s(<nil>)

=== RUN   TestPosixAppendFile
--- FAIL: TestPosixAppendFile (0.00s)
        posix_test.go:1388: Case: 10, expected: %!s(<nil>), got: open /tmp/minio-099676994/success-vol/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003/object000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001: file name too long
        posix_test.go:1414: expected: Permission error, got: %!s(<nil>)
=== RUN   TestPosixPrepareFile
--- FAIL: TestPosixPrepareFile (0.00s)
        posix_test.go:1489: Case: 7, expected: %!s(<nil>), got: open /tmp/minio-522472388/success-vol/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002/level0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000003/object000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001: file name too long
        posix_test.go:1515: expected: Permission error, got: %!s(<nil>)

=== RUN   TestServerSuite/Test:_5,_ServerType:_XLSet
--- FAIL: TestServerSuite (1.41s)
    --- FAIL: TestServerSuite/Test:_1,_ServerType:_FS (0.03s)
        server_test.go:61: Test [server_test.go:453:(*TestSuiteCommon).TestDeleteBucketNotEmpty()]:FS expected 409, got 500
    --- FAIL: TestServerSuite/Test:_2,_ServerType:_FS (0.02s)
        server_test.go:61: Test [server_test.go:453:(*TestSuiteCommon).TestDeleteBucketNotEmpty()]:FS expected 409, got 500
    --- FAIL: TestServerSuite/Test:_3,_ServerType:_FS (0.68s)
        server_test.go:61: Test [server_test.go:453:(*TestSuiteCommon).TestDeleteBucketNotEmpty()]:FS expected 409, got 500
    --- FAIL: TestServerSuite/Test:_4,_ServerType:_XL (0.08s)
        server_test.go:61: Test [server_test.go:453:(*TestSuiteCommon).TestDeleteBucketNotEmpty()]:XL expected 409, got 503
    --- FAIL: TestServerSuite/Test:_5,_ServerType:_XLSet (0.60s)
        server_test.go:61: Test [server_test.go:453:(*TestSuiteCommon).TestDeleteBucketNotEmpty()]:XLSet expected 409, got 503

=== RUN   TestWebHandlerDeleteBucket
--- FAIL: TestWebHandlerDeleteBucket (0.03s)
        web-handlers_test.go:381: expected `..Bucket not empty..` but got `remove /tmp/minio-457750290/gt5sm4n3r4aa6rs6k0552s9sdkxs3x2au5bjqqhhwdz4gtxyxngoii0adgkf: file exists`
        web-handlers_test.go:381: expected `..Bucket not empty..` but got `Multiple disks failures, unable to write data.`

=== RUN   TestCertNew
--- FAIL: TestCertNew (0.00s)
        certs_test.go:48: lstat server.crt: no such file or directory
=== RUN   TestValidPairAfterWrite
--- FAIL: TestValidPairAfterWrite (0.00s)
        certs_test.go:72: open server2.crt: no such file or directory
=== RUN   TestStop
--- FAIL: TestStop (0.00s)
        certs_test.go:101: open server2.crt: no such file or directory
FAIL
FAIL    github.com/minio/minio/pkg/certs        0.013s
?       github.com/minio/minio/pkg/cgroup       [no test files]
?       github.com/minio/minio/pkg/disk [no test files]
?       github.com/minio/minio/pkg/dns  [no test files]

=== RUN   TestLockAndUnlock
--- FAIL: TestLockAndUnlock (1.00s)
        lock_test.go:176: unexpected unblocking
        lock_test.go:189: unexpected blocking
FAIL
FAIL    github.com/minio/minio/pkg/lock 1.016s

=== RUN   TestGetStats
--- FAIL: TestGetStats (0.00s)
        stats_test.go:25: Tests: Expected `nil`, Got Not implemented
        stats_test.go:28: Tests: Expected `n > 0`, Got 0
FAIL
FAIL    github.com/minio/minio/pkg/sys  0.013s

```

could you please let me know how to try fix it and how to run tests partly where they are failed?

**Comment by ikozhukhov** [comment]
additional issues:
```
vendor/github.com/rjeczalik/notify/event_fen.go:24:15: undefined: fileAccess
vendor/github.com/rjeczalik/notify/event_fen.go:26:17: undefined: fileModified
vendor/github.com/rjeczalik/notify/event_fen.go:29:15: undefined: fileAttrib
vendor/github.com/rjeczalik/notify/event_fen.go:31:15: undefined: fileDelete
vendor/github.com/rjeczalik/notify/event_fen.go:33:17: undefined: fileRenameTo
vendor/github.com/rjeczalik/notify/event_fen.go:35:19: undefined: fileRenameFrom
vendor/github.com/rjeczalik/notify/event_fen.go:37:14: undefined: fileTrunc
vendor/github.com/rjeczalik/notify/event_fen.go:39:17: undefined: fileNoFollow
vendor/github.com/rjeczalik/notify/event_fen.go:41:14: undefined: unmounted
vendor/github.com/rjeczalik/notify/event_fen.go:43:16: undefined: mountedOver
vendor/github.com/rjeczalik/notify/event_fen.go:43:16: too many errors
Makefile:75: recipe for target 'build' failed
gmake: *** [build] Error 2
```

**Comment by nitisht** [comment]
Fixed by #6770

**Comment by ikozhukhov** [comment]
it fixed build, but not fixed tests what i posted before.
how we can do with failed tests - open another ticket per failed test or open one ticket for all failed tests ?

**Comment by harshavardhana** [comment]
> it fixed build, but not fixed tests what I posted before.
> how we can do with failed tests - open another ticket per failed test or open one ticket for all failed tests ?

That should be it @ikozhukhov - we have really no plans on supporting Solaris by making sure all tests are working. If there is an external contributor such as you who can maintain it then it is not a problem and we are more than happy to accept patches.  Most of our testing and users are on either Linux, Windows or MacOS.

So opening issue might be okay if you wish to track it here, but we do not plan to expend more effort beyond fixing build issues.
